### DISCUSSION TITLE
Watchdog Timer on attiny1614

### DISCUSSION THREAD

**Discussion by sfcdaddona** [SpenceKonde/megaTinyCore#493] [Q&A]
First off, Thanks Dr. Azzy for this awesome core. I am a novice, and this is my first time I can't work something out on my own.

I have been trying to implement the atitny 1614 native watchdog timer for a week, and I am having no luck.
I would like to perform a reset if program control does not go back to the main loop within 8 seconds. I would like to use normal, and not the windowed mode, which I understand is, windowed mode would do a hard reset if program control returned to the main loop to soon. 
I seem to have the below code working to reset every 8 seconds, but I can’t seem to find the correct command to reset the watchdog timer itself.
The 1614 datasheet references a WDR but I can't find out how to execute it.
Apologies for the laymen speak, I am new to this at an old age, and any help would be appreciated. 

Thanks,
Miles

cut from setup;
`  digitalWrite(IRLED, HIGH);delay(250);
  digitalWrite(IRLED, LOW);delay(200);
  digitalWrite(IRLED, HIGH); 
  LEDTimerStart = millis();
  **_PROTECTED_WRITE(WDT.CTRLA,0xB); // added and changed SpenceKonde readme.md, I thinks its normal mode 8secs**
    
}

void loop()//====================== Loop BEGIN ========================================//
{

**// Can't find the correct code to put here to reset the the Watchdog timer only (WDR)**

  CheckButtonState();
  CheckForAndSend();
  BTN_OneAndThree();
  
  // Turns off LED
  if ((millis() - LEDTimerStart) > LEDTimeout) {
    digitalWrite(IRLED, LOW);
  }
  // Turns off FCU after 60 min
  if ((millis() - LEDTimerStart) > ShutDownTimeOut) {
    digitalWrite(PowerEnable, LOW);
  }
}

//----------End Of Void Loop----------------------------//

void CheckButtonState()
{
  CB1 = digitalRead(SW1);
  CB2 = digitalRead(SW2);
  Volts = analogRead(VoltsP);

  if (CB1 == LOW) {
    setIRLED(); LEDTimerStart = millis();
    SW1State = SW1State + 1;
  }
  if (CB1 == HIGH) {
    SW1State = SW1State - 1;
  }
  if (SW1State >= OnDebounce) {
    SW1State = OnDebounce;
  }
  if (SW1State <= OffDebounce) {
    SW1State = OffDebounce;
  }
  if (CB2 == LOW) {
    setIRLED(); LEDTimerStart = millis();
    SW2State = SW2State + 1;
  }
  if (CB2 == HIGH) {
    SW2State = SW2State - 1;
  }
  if (SW2State >= OnDebounce) {
    SW2State = OnDebounce;
  }
  if (SW2State <= OffDebounce) {
    SW2State = OffDebounce;
  }

  if (Volts > 75) {
    setIRLED(); LEDTimerStart = millis();
    SW3State = SW3State + 1;
  }
 if (Volts <= 75) {
    SW3State = SW3State - 1;
  }
  if (SW3State >= OnDebounce) {
    SW3State = OnDebounce;
  }
  if (SW3State <= OffDebounce) {
    SW3State = OffDebounce;
  }
}`

TLTR
I am an amateur when it comes to coding micro-controllers, but I am really stuck on this one. I am using the 1614 as part of a wireless HID mouse and keyboard. The 1614 just does some IR emitter functions for a tracker and sends button states, battery voltage, ect to a 32u4 via NRF 24L01. Everything worked perfect for years using the 328PBs
Unfortunately now that I switched to the 1614, I get a very infrequent random lock up that requires a reset.  

Thanks again,

Miles

**Comment by sfcdaddona** [comment]
Here is the whole file if it helps.
[FCUv2_v300_072121_wdt_test.txt](https://github.com/SpenceKonde/megaTinyCore/files/6871939/FCUv2_v300_072121_wdt_test.txt)

**Comment by SpenceKonde** [comment]
1. Fix the comment on like 246 - don't write minimum timeout when you['re setting it to the maximum Don't change the code you copy-pasted without the changing the comments you pasted with it. 

2. I see no attempt to execute the watchdog reset there, so yeah, it'll keep resetting. `#include <avr/wdt.h>` and call `wdt_reset()` like on the classic parts (I don't recall off hand whether any other part of the wdt.h api is intact, I don't think so - but I also think they're all macros so the ones that don't work don't cause compile errors unless you try to use them), or define your own wdt_reset() - it has to execute the `wdr` instruction. This is impossible to do in pure C, because the C language has no concept of platform specific stuff like that (this is what most of the avr/something-not-starting-in-io.h headers do - wrap hardware functionality implemented as inline assembly in a macro so you can call it from C. Some are trivial, like this, others not entirely trivial, but I don't recall seeing any that are sophisticated. The peak of it's complexity I think is boot.h)
```c
// As a function
void wdt_reset() {
  __asm__ __volatile__ ("wdr"::);
}
```
```c
// as a macro (which is all that wdt.h does)
#define wdt_reset() __asm__ __volatile__ ("wdr"::)
```

Both produce binary identical output; and no matter how you get a wdt_reset() you call it like you would any function or functionlike macro. 

... And I did end up checking anyway. I think the enable/disable functions from wdt.h should work too. Not because AVR_libc libraries were updated for these parts or anything, but just because it's the same as the xmega parts for which it was updated ages ago.

You can lose the stupid double underscores on either side of the asm and volatile keywords, but if I go recommending that some standards weenie on a non-Arduino-IDE platform who is using avr-gcc with a more stringent mode will cry foul (those keywords, in that context, are not ISO standard C, you see, and compilers are supposed to prefix and postfix any beyond-the-standard keywords with that).

**Reply by sfcdaddona** [reply]
Wow, thank you so much for the quick reply, I am digesting it now.

Thanks again Sir.
Miles

**Comment by sfcdaddona** [comment]
I can't believe how quickly you solved this for me. The macro seemed to work! I searched the internet for days, and continued to reread the 1614 datasheet, only to continue to fail. I am looking forward to a long term test to see if the micro-controller locks up again.
I honestly don't know how you have the time develop what you do for everyone, and still help people like me, but I truly appreciate it.

You are a true selfless gentleman,

Miles

**Comment by SpenceKonde** [comment]
Lol - simple, I don't have the time to do all this stuff! 

As for this specific matter, IMO it's what support people call a gotcha, where the nature of the problem is immediately apparent amnd tjhe so;lution is known. 

Since I don't consider writing inline assembly to bne a particularly big deal, have used thw WDT on a bunch of parts, and and read the avr-libc headers for one reason or another on at least o9ne occsion per file.... it's not like I had to think tio solve it. I am addingn to the relevant section of the core docs, because I hadn't realized I hadnt' area,ly mnentioned how to actuazlly, like, se it from within a sketcb

**Comment by GordyWairoa** [comment]
Many thanks to this thread. 
I was having trouble figuring out how to set up the ATTINY1624 WDT.
I was beginning to try and direct setup the WDT period register and found the perfect solution here. 
Here is my test code for anyone else to see my simple test:

#include <avr/wdt.h>
/*
Watchdog Timer tests ATtiny1624
20240103
*/
void setup() {
//
// need to direct control the ATTINY1624 watchdog timer period register
_PROTECTED_WRITE(WDT.CTRLA,0x8); // now 1 sec
// Options 8MS, 16MS, 32MS, 64MS, 128MS,  256MS, 512MS, 1S,  2S,  4S,  8S
//         0x1, 0x2,  0x3,  0x4,  0x5, 0x6,  0x7, 0x8, 0x9, 0xA, 0xB
//
// initialize digital pin for LED as an output.
  pinMode(1, OUTPUT);
  digitalWrite(1, HIGH);  // turn the LED on
  delay(200);             // wait for a 200ms
  digitalWrite(1, LOW);   // turn the LED off
  delay(200);
//
}
//
void loop() {
wdt_reset();
// delay(100); // setup LED should not flash with this short delay
delay(2000); // setup LED should flash with this long delay
//
}

**Reply by SpenceKonde** [reply]
glad to be of assistance. 
BTW, you know that most of wdt.h doesn't work right? I assume you noticed by now.
